### PR TITLE
refactor(RELEASE-2249): split add-fbc-contribution into discrete steps

### PR DESCRIPTION
**Pull request by happybhati**
Split the monolithic add-contribution script (~520 lines) into three focused steps for improved readability, maintainability, and debuggability:

- prepare-inputs: validates inputs, groups components by OCP version, writes shared config for subsequent steps
- process-ocp-groups: executes IIB requests per OCP group with batching, index chaining, and retries
- collect-and-deduplicate: aggregates results, deduplicates by target, produces final output

No breaking changes - all task parameters, results, and volumes are preserved. The external contract with the fbc-release pipeline and downstream tasks remains identical.

I hope this helps with reviewing this change. Let me know of any suggestions or improvements after reviewing, thanks :)
[demo](https://drive.google.com/file/d/17c_UPPAk6_iyncTBtr5UAkHPzBNiULdy/view?usp=drivesdk)
[Miro board](https://miro.com/app/board/uXjVG4AoBNI=/?share_link_id=620329272779)

Assisted-by: Cursor AI

## Checklist before requesting a review
- [ ] I have marked as draft or added `do not merge` label if there's a dependency PR
  - If you want reviews on your draft PR, you can add reviewers or add the `release-service-maintainers` handle if you are unsure who to tag
- [ ] My commit message includes `Signed-off-by: My name <email>`
- [ ] I read CONTRIBUTING.MD and [commit formatting](CONTRIBUTING.md#commit-message-formatting-and-standards)
- [ ] I have run the README.md generator script in `.github/scripts/readme_generator.sh` and verified the results using `.github/scripts/check_readme.sh`

### CODE REVIEW

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml [hunk 1]
```diff
@@ -197,157 +205,325 @@ spec:
...
-        TASK_LABEL="internal-services.appstudio.openshift.io/group-id"
-        TASK_ID=$(context.taskRun.uid)
-        PIPELINERUN_LABEL="internal-services.appstudio.openshift.io/pipelinerun-uid"
-
-        # Create batches with size limits
-        LENGTH=$(jq -r '.components | length' "${SNAPSHOT_PATH}")
+        # --- Batch and Timeout Calculation ---
         MAX_BATCH_SIZE="$(params.maxBatchSize)"
-        echo "Processing $LENGTH components with maximum batch size of $MAX_BATCH_SIZE..."
+        echo "Processing $total_components components with maximum batch size of $MAX_BATCH_SIZE..."
```
Comment (theflockers) [inline tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml]: From my understanding, this part is just preparing/gathering variables and so on to later actually process, so maybe  the message "Processing something" might give a wrong impression that the batch processing actually started here.

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml [hunk 2]
```diff
@@ -197,157 +205,325 @@ spec:
...
 
-        # Calculate number of batches needed
-        NUM_BATCHES=$(( (LENGTH + MAX_BATCH_SIZE - 1) / MAX_BATCH_SIZE ))
-        echo "Creating $NUM_BATCHES batch(es) for $LENGTH components"
+        finally_task_timeout=300
+        pipeline_timeout=$(date -u "+%Hh%Mm%Ss" -d @$(( request_timeout_seconds + finally_task_timeout )))
+        task_timeout=$(date -u "+%Hh%Mm%Ss" -d @$(( request_timeout_seconds )))
 
+        # --- Group Components by OCP Version ---
         echo "INFO: Processing snapshot components"
```
Comment (theflockers) [inline tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml]: Same here. Maybe we should say something similar to "analyzing snapshot components to create the batch processing configuration".

**Comment by happybhati** [comment]
@chandwanitulsi and @yashvardhannanavati  I have created an introductory demo, which might help. Please refer to the Miro board diagram instead of the demo recording. The fonts are smaller in the recording, so please use Miro. Each diagram in Miro also has a caption if you want to read about a particular step or flow.

I hope this helps with reviewing this change. Let me know of any suggestions or improvements after reviewing, thanks :) 
[demo](https://drive.google.com/file/d/17c_UPPAk6_iyncTBtr5UAkHPzBNiULdy/view?usp=drivesdk)
[Miro board](https://miro.com/app/board/uXjVG4AoBNI=/?share_link_id=620329272779)

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml (hunk 1 above)
Comment (happybhati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml]: Good catch — updated all three "Processing" messages in the prepare step to reflect it's only gathering/grouping, not executing

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml (hunk 2 above)
Comment (happybhati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml]: fixed, thanks

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml [hunk 3]
```diff
@@ -160,35 +165,38 @@ spec:
...
-        # The snapshot should be updated by prepare-fbc-snapshot
         SNAPSHOT_PATH="$(params.dataDir)/$(params.snapshotPath)"
         if [ ! -f "$SNAPSHOT_PATH" ]; then
             echo "ERROR: Snapshot not found at $SNAPSHOT_PATH"
             exit 1
         fi
 
-        # adding a new result as modifying the current one used breaks e2e for single component.
+        # --- Write Tekton Result Paths ---
+        # Note: adding a new result as modifying the current one used breaks e2e for single component.
```
Comment (theflockers) [inline tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:195]: Maybe we should also take opportunity of this PR to refactor and also adjust this condition here. The result expected was necessary for single component mode, but now even single component is handled my multi-component routine, so this result probably is not necessary anymore.

Comment (happybhati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:195]: Good point — you're right that requestResultsFile is a leftover from the single-component path. I checked and extract-index-image actually reads from internalRequestResultsFile now, not inputDataFile. So the result and the param are both candidates for cleanup. But since that change touches extract-index-image and the pipeline YAML too, I'd prefer to handle it in a follow-up (maybe RELEASE-1640 which the comment already references) to keep this PR scoped to the refactoring. WDYT?

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml [hunk 4]
```diff
@@ -160,35 +165,38 @@ spec:
...
       computeResources:
         limits:
-          memory: 512Mi
-          cpu: 200m
+          memory: 256Mi
+          cpu: 100m
         requests:
-          memory: 512Mi  # was exiting with code 137 when set to 256Mi
-          cpu: 200m
+          memory: 256Mi
```
Comment (yashvardhannanavati) [inline tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:171]: should this be set to 512Mi instead? basing on the comment on line 170

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml [hunk 5]
```diff
@@ -197,157 +205,325 @@ spec:
...
-        # Snapshot validation for multi-OCP processing
+        # --- Snapshot Validation ---
         if ! jq -e '.components | type == "array" and length > 0' "$SNAPSHOT_PATH" > /dev/null; then
             echo "ERROR: Snapshot missing required 'components' array"
             exit 1
         fi
 
-        # Validate snapshot for processing
         echo "INFO: Validating snapshot for processing..."
         total_components=$(jq '.components | length' "$SNAPSHOT_PATH")
```
Comment (yashvardhannanavati) [inline tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:243]: Consider changing this to `jq '(.components | length) // 0' "$SNAPSHOT_PATH"` to avoid failing on line 251 in-case the value is null

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml [hunk 6]
```diff
@@ -367,29 +543,29 @@ spec:
...
-                if [ "${mustOverwriteFromIndexImage}" = "true" ]; then
-                    # IIB overwrites fromIndex - keep using original consistently
+                if [ "${must_overwrite_from_index_image}" = "true" ]; then
+                    # IIB overwrites fromIndex — keep using original consistently
                     printf '%s' "${group_from_index}"
                 else
-                    # IIB creates new index - we MUST track what was actually created
+                    # IIB creates new index — track what was actually created
                     if [ ${#group_successful_batches[@]} -eq 0 ]; then
-                        # No successful batches yet - start from original fromIndex
```
Comment (yashvardhannanavati) [inline tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:401]: it might be worth keeping this comment?

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml [hunk 7]
```diff
@@ -399,150 +575,47 @@ spec:
...
-                    return 1
+                if [ "${must_overwrite_from_index_image}" = "false" ]; then
+                    local result_file
+                    result_file="$(params.dataDir)/ir-$(context.taskRun.uid)"
+                    result_file+="-batch-$((batch_num + 1))-result.json"
+                    local batch_results
+                    batch_results=$(cat "${result_file}")
+                    group_latest_iib_index_image=$(jq -r '.jsonBuildInfo' <<< "${batch_results}" | \
+                        base64 -d | gunzip | jq -r '.index_image')
+                    echo "Updated fromIndex for next batch: ${group_latest_iib_index_image}"
```
Comment (yashvardhannanavati) [inline tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml]: Consider returning the value via stdout rather than updating the `group_latest_iib_index_image` variable inside the function.

**Review by yashvardhannanavati — COMMENTED** [review]
On a separate note, the existing tests do not seem to test batching. Everything is processed in a single batch. Maybe consider adding a test for that? 

Otherwise, only minor comments. It looks good.

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml (hunk 4 above)
Comment (theflockers) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:171]: In think the reduction is because the task was split in 3 steps, so in theory it does not need the full memory for the step.

Comment (happybhati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:171]: Leandro's right — since we split the task, this step only does input validation and jq grouping, so 256Mi is enough. The heavier step (process-ocp-groups at line 332) has 512Mi.

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml (hunk 5 above)
Comment (happybhati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:243]: Good thought, but line 244 already validates that `.components` is a `non-null` array before we reach line 250 — if it's null we `exit 1` at line 246. So this can't fail here. Happy to add the `// 0` if you'd prefer belt-and-suspenders though.

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml (hunk 6 above)
Comment (happybhati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:401]: Good eye — restored the comment. It was accidentally dropped during the refactoring.

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml (hunk 7 above)
Comment (happybhati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml]: Agreed with the principle — stdout return is the dominant pattern in the repo and the other helpers in this task use it. However, this function has a conditional guard (must_overwrite_from_index_image) where it intentionally produces no output, so a naive var=$(fn) would blank the variable. Making it safe requires a null-check wrapper at both call sites. Given the function is called in only 2 places and sets exactly 1 variable, keeping the direct mutation avoids introducing a subtle edge case. Happy to refactor if you feel strongly about it though.

**Comment by happybhati** [comment]
> On a separate note, the existing tests do not seem to test batching. Everything is processed in a single batch. Maybe consider adding a test for that?
> 
> Otherwise, only minor comments. It looks good.

Good catch — added test-add-fbc-contribution-multi-batch.yaml that uses maxBatchSize=1 with 2 components to force 2 batches. Verifies batch chaining, deduplication, and correct fragment survival.

**Comment by chandwanitulsi** [comment]
cc @nmars

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml (hunk 4 above)
Comment (yashvardhannanavati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:171]: ack, thank you

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml (hunk 5 above)
Comment (yashvardhannanavati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:243]: up to you. I don't feel strongly for or against it

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml (hunk 7 above)
Comment (yashvardhannanavati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml]: While we are avoiding the edge case right now, a future refactor might expose it and result in a bug. I prefer being explicit and having null-check wrapper at both call sites and returning the variable. I'll leave the final call on you though :)

Comment (happybhati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml]: Good call — refactored to stdout return with null-check at both call sites. Also renamed to get_index_from_batch_result for consistency with the other helpers. Tested both paths on Kind, all good.

**Comment by happybhati** [comment]
@theflockers @yashvardhannanavati could you please review again? Thanks 
Once I get approval, I will proceed with the architecture reviews.

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml [hunk 8]
```diff
@@ -195,157 +203,324 @@ spec:
...
 
-            echo "INFO: Processing completed successfully"
+            # Store raw results for this batch
+            echo "${results}" > \
+                "$(params.dataDir)/ir-$(context.taskRun.uid)-batch-$((batch_num + 1))-result.json"
+
+            # Decompress build info for fragment processing
+            local decompressed_json_build_info
+            decompressed_json_build_info="$(jq -r '.jsonBuildInfo' <<< "${results}" \
+                | base64 -d | gunzip)"
```
[the remark this answers is not in the document]
Comment (happybhati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:457]: The exitCode check a few lines above already catches failed InternalRequests — if jsonBuildInfo were missing, exitCode would be non-zero and we'd exit before reaching the decode.

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml [hunk 9]
```diff
@@ -195,157 +203,324 @@ spec:
...
         # Initialize results file for multi-OCP processing
-        echo -n "$timestamp" > "$(results.buildTimestamp.path)"
         jq -n '{"components": []}' | tee "$RESULTS_FILE"
 
-        # Snapshot validation for multi-OCP processing
+        # --- Snapshot Validation ---
         if ! jq -e '.components | type == "array" and length > 0' "$SNAPSHOT_PATH" > /dev/null; then
             echo "ERROR: Snapshot missing required 'components' array"
             exit 1
         fi
```
[the remark this answers is not in the document]
Comment (happybhati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:240]: The snapshot structure is already validated by prepare-fbc-snapshot upstream; if components array were missing the pipeline would have failed before reaching this task.

**Review by theflockers — DISMISSED** [review]
LGTM. Once the repo of the opt-in component is onboard/changed in pyxis, we re-test is and it is ready for architecture review.

**Comment by yashvardhannanavati** [comment]
as discussed with @happybhati and @theflockers, let's hold off on merging this until #2132 is merged. thank you!

**Comment by happybhati** [comment]
https://github.com/konflux-ci/release-service-catalog/pull/2132 is merged 
Resolved the conflicts and rebased, ready for review again.

CC: @yashvardhannanavati  and @theflockers 🙇

**Comment by happybhati** [comment]
@swickersh and @theflockers Could you please take a look whenever you get a chance, Thanks

**Comment by swickersh** [comment]
Hi @happybhati 
First off, nice demo and this change looks like a good split into steps with logical separation of concerns.
Is this task being converted to Python soon? if so - what's the best approach here?

I see you initially started this in February, so the Python migration wasn't on our radar as closely then. And obviously you had to wait for #2132 to land.

Once this PR get's architect approval - should we just jump right into converting it to python? Or maybe convert to Python now based off this and we all review that one instead?

Apologies if this has already been discussed and I missed it.

**Comment by happybhati** [comment]
> Hi @happybhati First off, nice demo and this change looks like a good split into steps with logical separation of concerns. Is this task being converted to Python soon? if so - what's the best approach here?
> 
> I see you initially started this in February, so the Python migration wasn't on our radar as closely then. And obviously you had to wait for #2132 to land.
> 
> Once this PR get's architect approval - should we just jump right into converting it to python? Or maybe convert to Python now based off this and we all review that one instead?
> 
> Apologies if this has already been discussed and I missed it.

Thanks for the review and kind words, @swickersh!

Yes, @elenagerman  has a draft PR for the Python conversion: [#869](https://github.com/konflux-ci/release-service-utils/pull/869) (RELEASE-2456). I saw [Johnny's comment ](https://github.com/konflux-ci/release-service-utils/pull/869#issuecomment-4961669179)there where he mentioned he'd prefer a single Python script since the organization with Python imports will be way cleaner than bash, and he's flexible on merge order as long as we end up with Python.

I'm open to both approaches:

1. Review and merge this bash refactor first, then convert to Python using it as a structural reference
2. Use this PR as a base for a direct Python conversion and review that instead

Either way, the logical separation (prepare-inputs, process-ocp-groups, collect-and-deduplicate) should carry over to the Python version.

**Comment by swickersh** [comment]
> > Hi @happybhati First off, nice demo and this change looks like a good split into steps with logical separation of concerns. Is this task being converted to Python soon? if so - what's the best approach here?
> > I see you initially started this in February, so the Python migration wasn't on our radar as closely then. And obviously you had to wait for #2132 to land.
> > Once this PR get's architect approval - should we just jump right into converting it to python? Or maybe convert to Python now based off this and we all review that one instead?
> > Apologies if this has already been discussed and I missed it.
> 
> Thanks for the review and kind words, @swickersh!
> 
> Yes, @elenagerman has a draft PR for the Python conversion: [#869](https://github.com/konflux-ci/release-service-utils/pull/869) ([RELEASE-2456](https://issues.redhat.com/browse/RELEASE-2456)). I saw [Johnny's comment ](https://github.com/konflux-ci/release-service-utils/pull/869#issuecomment-4961669179)there where he mentioned he'd prefer a single Python script since the organization with Python imports will be way cleaner than bash, and he's flexible on merge order as long as we end up with Python.
> 
> I'm open to both approaches:
> 
> 1. Review and merge this bash refactor first, then convert to Python using it as a structural reference
> 2. Use this PR as a base for a direct Python conversion and review that instead
> 
> Either way, the logical separation (prepare-inputs, process-ocp-groups, collect-and-deduplicate) should carry over to the Python version.

Gotcha, thanks for the details. Well this lgtm either way.
Both approaches you suggest are good with me. #2 makes a little more sense to me but I don't feel strongly about it. 

[RELEASE-2456]: https://redhat.atlassian.net/browse/RELEASE-2456?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Review by yashvardhannanavati — APPROVED** [review]

**Review by theflockers — DISMISSED** [review]
LGTM

**Review by johnbieren — DISMISSED** [review]
the demo was very helpful, thank you for creating it

**Comment by ach912** [comment]
## Production Approval Record

| Field | Value |
|---|---|
| **Action** | APPROVED |
| **Reviewer** | @ach912 |
| **Timestamp** | 2026-07-30T09:55:21.839Z |
| **Criteria Override** | AI review findings not addressed |

Approved

File: tasks/managed/add-fbc-contribution/tests/test-add-fbc-contribution-multi-batch.yaml [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+            - name: IMAGE_EXPIRES_AFTER
+              value: $(params.ociArtifactExpiresAfter)
+            - name: "ORAS_OPTIONS"
+              value: "$(params.orasOptions)"
+            - name: "DEBUG"
+              value: "$(params.trustedArtifactsDebug)"
+        steps:
+          - name: setup-test-data
+            image: quay.io/konflux-ci/release-service-utils@sha256:5546fa78d3c88d7b6a2e8cff8902f7757f00541d0bbaf113b9f293133894afa3
+            script: |
```
[the remark this answers is not in the document]
Comment (happybhati) [reply tasks/managed/add-fbc-contribution/tests/test-add-fbc-contribution-multi-batch.yaml:68]: Test pipeline inline steps don't require compute resources -- this is consistent with other test pipelines in the repo.

File: tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml [hunk 11]
```diff
@@ -193,177 +201,371 @@ spec:
...
-        mustOverwriteFromIndexImage="$(params.mustOverwriteFromIndexImage)"
+        must_publish_index_image="$(params.mustPublishIndexImage)"
+        must_overwrite_from_index_image="$(params.mustOverwriteFromIndexImage)"
 
-        # Extract OCP versions from snapshot (ocpVersion is JSON array)
+        # --- Extract and Validate OCP Versions ---
         # Components can have multiple versions: ["v4.17", "v4.18", "v4.19"]
         # Fallback to empty array if ocpVersion is absent or null for backward compatibility
         ocp_versions=$(jq -r '.components[] | (.ocpVersion // []) | .[]' "$SNAPSHOT_PATH" | sort -u | tr '\n' ' ')
         echo "INFO: Found OCP versions: ${ocp_versions}"
```
[the remark this answers is not in the document]
Comment (happybhati) [reply tasks/managed/add-fbc-contribution/add-fbc-contribution.yaml:226]: `prepare-fbc-snapshot` upstream already validates and hard-fails on missing `ocpVersion`, so this field is guaranteed by the time `add-fbc-contribution` runs. The `// []` is just a defensive fallback.

**Comment by happybhati** [comment]
PR #2329 (fix FBC manifest unknown) merged and caused two conflicts in add-fbc-contribution.yaml. Resolved and incorporated Leandro's changes into our refactored structure:

No logic changes beyond integrating #2329. Could I get a re-review please? 🙏

CC: @swickersh @theflockers @yashvardhannanavati @johnbieren @ach912 @davidmogar and @twaugh

**Review by johnbieren — APPROVED** [review]

**Review by ach912 — APPROVED** [review]

**Review by theflockers — APPROVED** [review]

**Comment by theflockers** [comment]
Amazing work on refactoring this, @happybhati . Thanks!
